nyc_taxi: Modernize notebooks

Azaya89 opened the pull request:
This PR modernizes the `nyc_taxi.ipynb`, `nyc_taxi-nongeo`, and `dashboard.ipynb` notebooks. After updating the dependencies however, I noticed a bug with the current holoviews version which was fixed [here](https://github.com/holoviz/holoviews/pull/6155). This will require updating the channels when the new holoviews version is released.

- [x] Tests passed.

---

# Modernizing an example checklist

## Preliminary checks

- [x] Look for [open PRs](https://github.com/holoviz-topics/examples/pulls) and [issues](https://github.com/holoviz-topics/examples/issues) that reference the project you are updating. It is possible previous unmerged work in PR could be re-used to modernize the project. Comment on these PRs and issues when appropriate, hopefully we should be able to close some of them after your modernizing work.

## Change ‘anaconda-project.yml’ to use the latest workable version of packages

- [x] Pin python=3.11
- [x] Remove the upper pin (e.g. `hvplot<0.9` to `hvplot`, `panel>=0.12,<1.0` to `panel>=0.12`) of all other dependencies. Removing the upper pins of dependencies could necessitate code revisions in the notebooks to address any errors encountered in the updated environment. Should complexities or extensive time requirements arise, document issues for team discussion on whether to re-pin specific packages or explore other solutions.
- [x] Add/update the lower pin of all other dependencies (e.g. `hvplot` to `hvplot>=0.9.2`, `hvplot>=0.8` to `hvplot>=0.9.2`). Usually, the new/updated lower pin of a dependency will be the version resolved after `anaconda prepare` has been run. Execute `!conda list` in a notebook, or `anaconda run conda list` in the terminal, to display the version of each dependency installed in the environment. Adjusting the lower pin helps ensure that the locks produced for each platform (linux-64, win-64, osx-64, osx-arm64) rely on the tested dependencies and not on some older versions.
- [x] If one of the channels include conda-forge or pyviz, ask Maxime if it can be removed

## Plot API updates (discussed on a per-example basis)

- [x] Generally, try to replace HoloViews usage with hvPlot. At a certain point of complexity, such as with the use of ‘.select’, it might be better to stick with HoloViews. Additional examples of ‘complexity boundaries’ should be documented in [this document](https://docs.google.com/document/d/1LYkChbVZzqq5S1S4Mkj1rM3PNPJDDGriks8YTc4EQc4/edit?usp=sharing).
- [x] Almost always, try to replace the use of `datashade` with `rasterize` (read [this page](https://holoviews.org/user_guide/Large_Data.html)). Essentially, `rasterize` allows Bokeh to handle the colormapping instead of Datashader.

## Interactivity API updates (discussed on a per-example basis)

- [x] Remove all `pn.interact` usage
- [x] Avoid `.param.watch()` usage. This is [pretty low-level and verbose approach](https://param.holoviz.org/user_guide/Dependencies_and_Watchers.html#watchers) and should not be used in Examples unless required, or an Example is specifically trying to demo its usage in an advanced workflow.
- [x] Prefer using `pn.bind()`. Read [this page](https://panel.holoviz.org/explanation/api/reactive.html) for explanation.
- [x] For apps built using a class approach, when they create a `view()` method and call it directly, update the class by inheriting from `pn.viewable.Viewer` and replace `view()` by `__panel__()`. Here is an [example](https://panel.holoviz.org/gallery/nyc_deckgl.html). 

## Panel App updates (discussed on a per-example basis)

- [x] If the project doesn’t at any point create a Panel app at all, consider creating one. It can be as simple as wrapping a plot in `pn.Column`, or more complicated to incorporate widgets, etc. Make the final app  `.servable()`.
- [x] If the project creates an app in a notebook but doesn’t deploy it (i.e. there is no `command: dashboard` declaration in the `anaconda-project.yml` file), try adding it.
- [x] If the project already deploys an app but doesn’t wrap it in a nice template, consider wrapping it in a [template](https://panel.holoviz.org/reference/index.html#templates).
- [x] If the project deploys an app wrapped in a template, customize the template a little so all the apps don’t look similar (e.g. change the header background color). This doesn’t need to be discussed.
- [x] Comment start If you are building the application in a single cell, you can construct a template explicitly, like `template = pn.template.BootstrampTemplate`, but if building up an app across multiple cells, it is probably cleaner to declare the template at the top with `pn.extension(template='bootstrap')`. See [how to guide on setting a template](https://panel.holoviz.org/how_to/templates/template_set.html).

## General code quality updates

- [x] If the notebook disables warnings (e.g. with `warnings.simplefilter(‘ignore’)` somewhere at the start of the notebook, remove this line. Try to update the code to remove the warnings, if any. If updating the code to remove the warnings is taking significant amount of time and effort, bring it up for discussion and we may decide to disable warnings again.

## Text content

- [x] Edit the text content anywhere and everywhere that it can be improved for clarity.
- [x] Check the links are valid, and update old links (e.g. http -> https, xyz.pyviz.org -> xyz.holoviz.org)
- [x] Remove instructions to install packages inside an example

## Visual appearance - Example

- [x] Check that the titles/headings make sense and are succinct.
- [x] Check that the text content blocks are easily readable; revise into additional paragraphs if needed.
- [x] Check that the code blocks are easily readable; revise as needed. (e.g. add spaces after commas in a list if there are none, wrap long lines, etc.)
- [x] Check image and plot sizes. If possible, making them responsive is highly recommended.
- [x] Check the appearance on a smartphone (check Google to see how to adapt the appearance of your browser to display pages as if they were seen from a smartphone, this is usually done via the web developer tools). This is not a top priority for all examples, but if there are a few easy and straightforward changes to make that can improve the experience, let’s do it.
- [x] Check the updated notebook with the original notebook

## Visual appearance - Gallery

- [x] Check the thumbnail is visually appealing
- [x] Check the project title is well formatted (e.g. `Ml Annotators` to `ML Annotators`), if not, add/update the `examples_config.title` field in `anaconda-project.yml`
- [x] Check the project description is appropriate, if not, update the `description` field in `anaconda-project.yml`

## Workflow (after you have made the changes above)

- [x] Run successfully `doit validate:<projectname>`
- [x] Run successfully `doit test:<projectname>`
- [x] Run successfully `doit doc_one –name <projectname>`. It’s better if the project notebook(s) is saved with its outputs (but be sure to clear outputs before committing to the examples repo!) when building the docs. Then open this file in your browser `./builtdocs/index.html` and check how the site looks.
- [x] If you’re happy with all the above, open a PR. Reminder, clear notebook outputs before pushing to the PR.

Hoxbro commented on nyc_taxi/anaconda-project.yml:
```diff
@@ -18,22 +18,22 @@ channels:
+- bokeh>=3.3.4
+- datashader
+- fastparquet
+- holoviews
```
```suggestion
- pyviz/label/dev::holoviews>=1.19.0a0
```

Try this.

Azaya89 replied:
It's giving an error when i try to re-lock the envs:

```
Channels:
 - pyviz
 - defaults
Platform: osx-arm64
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... failed

PackagesNotFoundError: The following packages are not available from current channels:

  - holoviews==1.19.0a0=py_0

Current channels:

  - https://conda.anaconda.org/pyviz
  - defaults

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.


missing requirement to run this project: The project needs a Conda environment containing all required packages.
  '/Users/mac/Documents/development/holoviz-topics-examples/nyc_taxi/envs/default' doesn't look like it contains a Conda environment yet.
'/Users/mac/Documents/development/holoviz-topics-examples/nyc_taxi/envs/default' doesn't look like it contains a Conda environment yet.
Locking failed, re-set old lock file
TaskError - taskid:lock:nyc_taxi
PythonAction Error
Traceback (most recent call last):
  File "/Users/mac/Documents/development/holoviz-topics-examples/dodo.py", line 863, in lock_project
    subprocess.run(
  File "/Users/mac/anaconda3/envs/examples-gallery-manage/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['anaconda-project', 'lock', '--directory', 'nyc_taxi']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mac/anaconda3/envs/examples-gallery-manage/lib/python3.10/site-packages/doit/action.py", line 461, in execute
    returned_value = self.py_callable(*self.args, **kwargs)
  File "/Users/mac/Documents/development/holoviz-topics-examples/dodo.py", line 875, in lock_project
    shutil.move(oldlockf, lockf)
  File "/Users/mac/anaconda3/envs/examples-gallery-manage/lib/python3.10/shutil.py", line 816, in move
    os.rename(src, real_dst)
TypeError: rename: src should be string, bytes or os.PathLike, not NoneType

```

Hoxbro replied:
It should use pyviz/label/dev and not pyviz. Try to update the main channel to it.

Azaya89 replied:
Thanks. It worked now.

jbednar commented:
Thanks! Can you tell me what steps will let me reproduce this running project? E.g. `anaconda-project prepare` followed by something to use the git master version of holoviews (or maybe make a dev release?)?

maximlt commented:
> Thanks! Can you tell me what steps will let me reproduce this running project? E.g. `anaconda-project prepare` followed by something to use the git master version of holoviews (or maybe make a dev release?)?

It's been updated to depend on `pyviz/label/dev`. One question though, didn't we say we wanted the projects to depend on `defaults` as much as possible? During the big update end of last year, I removed pyviz and conda-forge as much as possible.

Azaya89 commented:
> Thanks! Can you tell me what steps will let me reproduce this running project? E.g. `anaconda-project prepare` followed by something to use the git master version of holoviews (or maybe make a dev release?)?

I think you can just run it the same way as before (i.e, `anaconda-project run`), since I've updated the lock files to use the dev version of holoviews.

> 
> It's been updated to depend on `pyviz/label/dev`. One question though, didn't we say we wanted the projects to depend on `defaults` as much as possible? During the big update end of last year, I removed pyviz and conda-forge as much as possible.

I think this is only temporary until the new holoviews version is released. Afterwards, we can update the lock files.
@Hoxbro

Hoxbro commented:
As @Azaya89 said, it is temporary until the next release of HoloViews is out.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -93,9 +90,7 @@
-    "scatter = hv.Scatter(samples, 'trip_distance', 'fare_amount')\n",
-    "labelled = scatter.redim.label(trip_distance=\"Distance, miles\", fare_amount=\"Fare, $\")\n",
-    "labelled.redim.range(trip_distance=(0, 20), fare_amount=(0,40)).opts(size=5)"
+    "samples.hvplot.scatter('trip_distance', 'fare_amount', xlabel='Distance, miles', ylabel='Fare, $', s=5, xlim=(0,20), ylim=(0,40))"
```
Just a very quick comment, we should try to avoid long line lengths as it means users will have to scroll horizontally which isn't great. That's what I see on my desktop screen, and it would be worse on a mobile phone or tablet.

![image](https://github.com/holoviz-topics/examples/assets/35924738/5745e21e-aa8c-4665-a988-e8ac01c17c39)

maximlt commented:
> As @Azaya89 said, it is temporary until the next release of HoloViews is out.

Ok noted thanks. Let's switch back to the `defaults` channel when HoloViews is released with this fix.

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -11,12 +11,12 @@
     "The [NYC Taxi trips](https://anaconda.org/jbednar/nyc_taxi/notebook) dataset is a well-studied data science example. Here we show how to build a simple dashboard for exploring 10 million taxi trips in a Jupyter notebook using [Datashader](http://datashader.org), then deploying it as a standalone dashboard using [Panel](http://panel.pyviz.org).\n",
     "\n",
-    "Running the dashboard requires having a live Python process running (not just a static webpage or anaconda.org notebook viewer).  Before you run the notebook or server, you'll need to set up a [conda](http://conda.pydata.org/miniconda.html) environment and run `conda install -c pyviz datashader panel holoviews`, and you'll need to [get the data file](https://anaconda.org/jbednar/nyc_taxi/not ...
+    "Running the dashboard requires having a live Python process running (not just a static webpage or anaconda.org notebook viewer).  Before you run the notebook or server, you'll need to set up a [conda](http://conda.pydata.org/miniconda.html) environment and run `conda install datashader panel holoviews`, and you'll need to [get the data file](https://anaconda.org/jbednar/nyc_taxi/notebook)."
```
I'd suggest removing this paragraph:
- `requires having a live Python process running` -> This applies to all the notebooks, not just this one.
- setup (conda + data download) -> We assume users have set up the project with anaconda-project

Azaya89 replied:
OK

maximlt replied:
Please remove this, as per the second item in my previous comment:

> Before you run the notebook or server, you'll need to set up a [conda](http://conda.pydata.org/miniconda.html) environment and run `conda install datashader panel holoviews`, and you'll need to [get the data file](https://anaconda.org/jbednar/nyc_taxi/notebook)."

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -11,12 +11,12 @@
+    "<img align='right' style='padding:0px 0px 0px 0px;' \n",
+    "src='./thumbnails/index.png'/>\n",
     "\n",
     "The [NYC Taxi trips](https://anaconda.org/jbednar/nyc_taxi/notebook) dataset is a well-studied data science example. Here we show how to build a simple dashboard for exploring 10 million taxi trips in a Jupyter notebook using [Datashader](http://datashader.org), then deploying it as a standalone dashboard using [Panel](http://panel.pyviz.org).\n",
```
- I don't think there's a need to link to this old notebook.
- Updating old links is useful, I'll add that to the checklist.

```suggestion
    "The NYC Taxi trips dataset is a well-studied data science example. Here we show how to build a simple dashboard for exploring 10 million taxi trips in a Jupyter notebook using [Datashader](https://datashader.org), then deploying it as a standalone dashboard using [Panel](https://panel.holoviz.org).\n",
```

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -33,19 +33,34 @@
-    "opts = dict(width=1000,height=600,xaxis=None,yaxis=None,bgcolor='black',show_grid=False)\n",
-    "cmaps = ['fire','bgy','bgyw','bmy','gray','kbc']\n",
-    "\n",
+    "hv.extension('bokeh', logo=False)"
```
No idea why `logo=False` was set, I don't think the logo hurts at all, let's keep it.

```suggestion
    "hv.extension('bokeh')"
```

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -11,12 +11,12 @@
-    "<img align='right' style='padding:30px 0px 40px 20px' \n",
-    "src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUAAAADkCAYAAADpav3mAAAgAElEQVR4Aey9eZBd53Un9rvr2/f3+r3X+w40NgLgJpKixF2S7diO18QT13jsZCaTVFKVqqTyR6omTmbsVKqy/OUkTmpqnHImTmk8Y0eW5UiiSIkiKRIECBBoAN0N9L6/fV/umjrnvtt4DTQgkGMllN0Xhb73ffd8yz33u+c7+yfUm01bgAA6bNhwr7kANtB3j64dyMOwVK//nlPXadGp3w9xr817cP1X9+674xHcThlM6BunC0vnw4d753Dp4/36bHVpkPeP46iyR4/hs/X96Dadu0eP5bP259Rz2nycNo6GOXpMj36a+9//0dBH93c ...
+    "<img align='right' style='padding:0px 0px 0px 0px;' \n",
+    "src='./thumbnails/index.png'/>\n",
```
I think this could be better displayed on the website, at the moment it makes the first two code cells quite small. Could you instead display the screenshot just below the first paragraph? Perhaps build the website locally for iterating on this change.

![image](https://github.com/holoviz-topics/examples/assets/35924738/5bbfa987-877b-420d-b9a7-0939f1ee48f7)

maximlt replied:
Yes I also tried, and it's more complicated than it should be!

It turns out our Sphinx `conf.py` adds the `html_image` myst extension (https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#html-images) that is used to let Sphinx know it should gather all the images it finds in `img` HTML tags.

> Allowed attributes are equivalent to the image directive: src, alt, class, width, height and name. Any other attributes will be dropped.

So the extension drops `styles` :( An alternative would be to use the myst syntax for images (https://myst-parser.readthedocs.io/en/latest/syntax/images_and_figures.html), but using that syntax means that it won't be displayed to users opening the notebook without the jupyterlab myst extension installed. Maybe, in the future, we'll add this extension to each project, but that will require more thought, so let's not do it now.

So my recommendation is to add this after the first paragraph, saving the image I shared below in `nyc_taxi/assets/dashboard.png`.

 ```
<img src='./assets/dashboard.png' alt="Panel Dashboard" align="center">
```

![dashboard](https://github.com/holoviz-topics/examples/assets/35924738/91346329-09c2-4752-9b5a-538a63e2a27d)

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -33,19 +33,34 @@
+   "source": [
+    "usecols = ['dropoff_x', 'dropoff_y', 'pickup_x', 'pickup_y', 'dropoff_hour', 'pickup_hour', 'passenger_count']\n",
+    "df = dd.read_parquet('data/nyc_taxi_wide.parq', engine='fastparquet')[usecols].persist()\n",
+    "opts = dict(width=1000, height=600, xaxis=None, yaxis=None, bgcolor='black', show_grid=False)\n",
```
Could you add a cell `df.head()` or `df.tail()` somewhere? I find it always useful to see a sample of the data in a notebook.

Azaya89 replied:
`df.head()` is already there. I'm surprised it didn't show up here.

maximlt replied:
Oops sorry I missed it!

maximlt commented on nyc_taxi/anaconda-project.yml:
```diff
@@ -14,26 +14,26 @@ examples_config:
+- pandas
+- panel
+- python-snappy
+- tornado
```
I don't think we need to specify `tornado`. I guess it had to be pinned at some point in time to work around an issue, that is likely no longer there.

Azaya89 replied:
I haven't sent the updated commits yet, but I've resolved them locally. Usually I mark each review as resolved as soon as I resolve the issues locally. This helps me work through the reviews systematically.

maximlt replied:
Ok, I see. I was confused because I think there were commits pushed after the review, so I thought this was missed. Your approach works for me.

maximlt commented on nyc_taxi/anaconda-project.yml:
```diff
@@ -14,26 +14,26 @@ examples_config:
+- panel
+- python-snappy
+- tornado
+- xarray
+- numba
```
Same comment for `xarray` and `numba`. They are not imported directly in the notebooks, and they are dependencies of `datashader` that is also listed.

maximlt replied:
Same comment as for `tornado`.

maximlt commented on nyc_taxi/anaconda-project.yml:
```diff
@@ -42,7 +42,7 @@ commands:
     supports_http_options: true
   notebook:
-    notebook: .
+    notebook: index.ipynb
```
I like that! In fact, I think that all the projects that an `index.ipynb` file should point to it.

maximlt commented on nyc_taxi/anaconda-project.yml:
```diff
@@ -14,26 +14,26 @@ examples_config:
+- pyviz/label/dev
 - defaults
 
 packages: &pkgs
```
- Before merging this PR, let's wait for HoloViews 1.19.0 to be released, and then remove `pyviz/label/dev` and update the lock file
- Please add/update the lower pins. After doing so, you'll have to re-lock the project, so I recommend you do it at together with the previous item. This step is described in the checklist:

> Add/update the lower pin of all other dependencies (e.g. `hvplot` to `hvplot>=0.9.2`, `hvplot>=0.8` to `hvplot>=0.9.2`). Usually, the new/updated lower pin of a dependency will be the version resolved after `anaconda prepare` has been run. Execute `!conda list` in a notebook, or `anaconda run conda list` in the terminal, to display the version of each dependency installed in the environment. Adjusting the lower pin helps ensure that the locks produced for each platform (linux-64, win-64, osx-64, osx-arm64) rely on the tested dependencies and not on some older versions.

Azaya89 replied:
OK. We wait!

maximlt commented on nyc_taxi/anaconda-project.yml:
This is how the project appears in the gallery:

![image](https://github.com/holoviz-topics/examples/assets/35924738/5acf837a-e97e-4241-95b6-f02a17ae79cc)

- I suggest its title should be `NYC Taxi` instead of `Nyc Taxi`. You can override the title automatically inferred from the project name by setting the `title` property in `examples_config`. Check the template project file.
- The thumbnail doesn't look so good to me, what do you think?
- Feel free to update the description if you think it could be improved.

Azaya89 replied:
I've changed the title but the thumbnail looks OK to me. It's basically the same as it was previously.

maximlt replied:
Ok, if you think it looks good, fine by me.

maximlt reviewed (changes requested):
I still have to review the geo and non-geo notebook changes but I already wanted to share what I had reviewed so far.

Azaya89 commented:
I'm getting this warning when I run `doit validate:nyc_taxi`: "WARNING: Notebook nyc_taxi/index.ipynb must not contain any code cell outputs, please clear it." but the `index.ipynb` file doesn't even have any code cell output in the first place. What could be the issue?

maximlt commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -428,9 +446,10 @@
+    "colormap = cc.CET_C1s\n",
+    "num_hours = 24\n",
+    "indices = np.linspace(0, len(colormap) - 1, num_hours, dtype=int)\n",
+    "colors = [colormap[i] for i in indices]"
```
<img width="1040" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/e588d473-684e-404e-a479-3b9fbbae0abf">

maximlt replied:
Maybe with `cc.CET_C7s`?

<img width="914" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/8eb6698b-549e-421d-a7ac-204da0e9df9e">

You can easily display the cyclic color maps with:

```python
from colorcet.plotting import swatches
import holoviews as hv
hv.extension('bokeh')

swatches(group=['cyclic'])
```

Azaya89 replied:
Thanks. I'll admit here that I'm not so good with color selection hence why I asked in the discord channel previously if the colors I selected were OK. I was more interested in using the colormaps available via `colorcet` rather than the old method of just listing the colors manually.

Azaya89 replied:
I've used `CET_C7s` now

maximlt commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -360,6 +369,8 @@
     "\n",
```
I think this code can be adapted to no longer use `shade()`. Please try this out:

```
def transform(img):
    agg = img.data["dropoff_x_dropoff_y Count"]
    return img.clone(agg.where(agg>np.percentile(agg, 90)))

custom_points = hv.Points(df, ['dropoff_x', 'dropoff_y'])
custom_rasterized = hd.rasterize(custom_points).apply(transform).opts(cmap='fire', cnorm='eq_hist', tools=['hover'])
tiles = EsriStreet().redim.range(x=x_range, y=y_range).opts(alpha=0.5)

tiles * hd.dynspread(custom_rasterized, threshold=0.3, max_px=4)
```

Azaya89 replied:
Nice!

maximlt commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -440,7 +459,7 @@
    "source": [
     "pickups = hv.Points(df, ['pickup_x', 'pickup_y'])\n",
-    "shade =  hd.datashade(pickups, aggregator=ds.count_cat('pickup_hour'), color_key=colors)\n",
+    "shade =  hd.datashade(pickups, aggregator=ds.by('pickup_hour'), color_key=colors)\n",
```
Out of curiosity, why updating from `ds.count_cat` to `ds.by`? I'm not super up to date on Datashader's API 🙃

Azaya89 replied:
It was recommended by Jim (March 20 meeting) as the more updated option in Datashader.

maximlt reviewed (commented):
I have reviewed the geo notebook.

maximlt commented on nyc_taxi/dashboard.ipynb:
```diff
@@ -11,12 +11,12 @@
+    "<img align='right' style='padding:0px 0px 0px 0px;' \n",
+    "src='./thumbnails/index.png'/>\n",
     "\n",
     "The [NYC Taxi trips](https://anaconda.org/jbednar/nyc_taxi/notebook) dataset is a well-studied data science example. Here we show how to build a simple dashboard for exploring 10 million taxi trips in a Jupyter notebook using [Datashader](http://datashader.org), then deploying it as a standalone dashboard using [Panel](http://panel.pyviz.org).\n",
```
Oops I was sure I commented that already, but no :) Could you:

- Remove the first link to Jim's old notebook, that doesn't look useful
- Update http to https links
- Update panel.pyviz.org to panel.holoviz.org

?

Azaya89 replied:
You made the comment initially and I've mostly resolved it.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -93,9 +89,8 @@
-    "scatter = hv.Scatter(samples, 'trip_distance', 'fare_amount')\n",
-    "labelled = scatter.redim.label(trip_distance=\"Distance, miles\", fare_amount=\"Fare, $\")\n",
-    "labelled.redim.range(trip_distance=(0, 20), fare_amount=(0,40)).opts(size=5)"
+    "samples.hvplot.scatter('trip_distance', 'fare_amount', xlabel='Distance, miles', ylabel='Fare, $',\n",
```
Could you wrap this line and others so they are a little less wide? So readers don't have to scroll too much when they're on their phone.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -134,18 +128,15 @@
-    "shaded = hd.spread(hd.datashade(ranged))\n",
-    "labelled = shaded.redim.label(trip_distance=\"Distance, miles\", fare_amount=\"Fare, $\")\n",
-    "labelled"
+    "df.hvplot.scatter('trip_distance', 'fare_amount', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
```
I don't think setting `alpha=0.05` is required here.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -134,18 +128,15 @@
-    "labelled = shaded.redim.label(trip_distance=\"Distance, miles\", fare_amount=\"Fare, $\")\n",
-    "labelled"
+    "df.hvplot.scatter('trip_distance', 'fare_amount', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
+    "                  xlabel='Distance, miles', ylabel='Fare, $', xlim=(0,15), ylim=(0,40), alpha=0.05)"
```
The plot looks quite different with these changes, the points on the new plot are quite small, they're a bit hard to see in the bottom part of the plot.

Before:
<img width="821" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/fd3d2fc6-d55b-41cd-869d-44e3f3f92851">

Now:
<img width="817" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/f0001370-0dfb-4019-bdd7-13169724debe">

I played around a little with it, it seems like it's mostly due to using `dynspread` instead of `spread`. The `spread` operation is not exposed to hvPlot users, there's I assume a good reason for that. Let's ask Jim's opinion.

Azaya89 replied:
I tried removing the `alpha` parameters and it became better, although still not as visible as the old one.

maximlt replied:
Ok, let's try to discuss this during our meeting tomorrow.

maximlt replied:
@Azaya89 can you please add these parameters to hvPlot calls that set `dynspread=True`? `threshold=1, max_px=1`.

Also, please make sure to arrange the parameters in the following order (when applicable): `rasterize`, `cnorm`, `dynspread`, `threshold`, `max_px`. Basically, let's put together parameters that are related. This applies to other parameters like xlim/ylim or xlabel/ylabel, I think we already do a good job for others but make sure it's the case.

<img width="823" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/0661cad8-c867-4b4a-9943-d65c8ec0f036">

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -154,9 +145,8 @@
-    "shaded = hd.spread(hd.datashade(ranged, normalization='linear'))\n",
-    "labelled = shaded.redim.label(trip_distance=\"Distance, miles\", fare_amount=\"Fare, $\")\n",
-    "labelled"
+    "df.hvplot.scatter('trip_distance', 'fare_amount', rasterize=True, dynspread=True,\n",
```
Same comment as above on `alpha` and on the fact that the plot looks quite different.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -172,11 +162,8 @@
-    "shaded = hd.spread(hd.datashade(ranged))\n",
-    "labelled = shaded.redim.label(tip_amount=\"Tip, $\", fare_amount=\"Fare, $\")\n",
-    "labelled"
+    "df.hvplot.scatter('tip_amount', 'fare_amount', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
```
This one is also a little difficult to see :/

<img width="825" alt="image" src="https://github.com/holoviz-topics/examples/assets/35924738/d35ea467-fb75-4947-9a83-ad074e267748">

maximlt replied:
Looking at this plot and reading the comment, I wish it displayed a diagonal (slope with gradient 1 and y-intercept 0) to make it obvious some tips are higher than fares. I opened an issue to record this idea https://github.com/holoviz/hvplot/issues/1309.

maximlt replied:
This is not going to be implemented soon in hvPlot so let's not hold off this PR for that.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -192,11 +179,8 @@
-    "shaded = hd.spread(hd.datashade(ranged, x_sampling=0.15))\n",
-    "labelled = shaded.redim.label(passenger_count=\"Passengers\", tip_amount=\"Tip, $\")\n",
-    "labelled"
+    "df.hvplot.scatter('passenger_count', 'tip_amount', rasterize=True, cnorm='log',\n",
```
Wow it's quite nice to see that this plot actually can be created with hvPlot! I would recommend also setting `y_sampling` (e.g. to `0.5`) so when you hover over the plot, you can see the number of people who tipped between [$x, $x+0.5]. If you make this update, please update the text below to indicate that.

Azaya89 replied:
I added `y_sampling` but there was no update to the hover info as far as I could tell.

maximlt replied:
Oh I didn't mean this would change the hover info. But by setting `y_sampling` you make the hover info more understandable, because now it's related to a known y-unit.

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -205,27 +189,7 @@
-    "Here datashader is aggregating over both fare_amount and trip_distance, but trip_distance was specified to have only a height of 2, because it will be further collapsed to create the histogram being displayed.  You can now use the wheel zoom tool when hovering over the x axis, and the plot will zoom in or out, dynamically resampling at the given location to make a new histogram (as long as t ...
-    "\n",
-    "In this particular plot, there is a very wide range of fare amounts, with an implausibly high maximum fare of over 4000 dollars, but you can easily zoom in to the bulk of the data to show that nearly all fares are between 4 and 20 dollars, following something like a gamma distribution, and they are discretized to the nearest 50 cents in this dataset."
+    "The above plots use the [Hvplot](https://hvplot.holoviz.org/) library, which builds Bokeh and Matplotlib plots from high-level specifications."
```
Plotly too :)

maximlt commented on nyc_taxi/nyc_taxi-nongeo.ipynb:
```diff
@@ -205,27 +189,7 @@
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "Here datashader is aggregating over both fare_amount and trip_distance, but trip_distance was specified to have only a height of 2, because it will be further collapsed to create the histogram being displayed.  You can now use the wheel zoom tool when hovering over the x axis, and the plot will zoom in or out, dynamically resampling at the given location to make a new histogram (as long as t ...
```
I assume you have agreed with Jim to remove this last example?

Azaya89 replied:
Yes.

maximlt reviewed (commented):
And I have re-viewed the last non-geo notebook! Left quite a lot of comments, but overall @Azaya89, great job!!! I see a lot of value in the modernizing work we're doing, it's really nice.

maximlt commented:
> I'm getting this warning when I run `doit validate:nyc_taxi`: "WARNING: Notebook nyc_taxi/index.ipynb must not contain any code cell outputs, please clear it." but the `index.ipynb` file doesn't even have any code cell output in the first place.
> 
> Does it have something to do with #377 ? @maximlt

Good catch thanks! Should be fixed by https://github.com/holoviz-topics/examples/pull/380.

maximlt commented:
I updated your original post with the modernizing checklist.

Azaya89 commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -80,9 +80,9 @@
     "import holoviews as hv\n",
     "from holoviews import opts\n",
-    "from holoviews.element.tiles import EsriStreet\n",
     "hv.extension('bokeh')"
```
> The line below with `plot_options = hv.Options(width=plot_width, height=plot_height, xaxis=None, yaxis=None)` looks like it does nothing at all. Try removing it please.

What line number exactly are you referring to?

maximlt replied:
This line: `plot_options = hv.Options(width=plot_width, height=plot_height, xaxis=None, yaxis=None)`.

Azaya89 commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -137,54 +135,47 @@
-    "Moreover, we're still ignoring 99% of the data.  Many plotting programs will have trouble with plots even this large, but Bokeh can handle 100-200,000 points in most browsers. Here we've enabled Bokeh's WebGL support, which gives smoother zooming behavior, but the non-WebGL mode also works well. Still, for such large sizes the plots become slow due to the large HTML file sizes involved, beca ...
+    "However, it's very tricky to set the size and alpha parameters.  How do we know if certain regions are saturating, unable to show peaks in dropoff density? Here we've manually set the alpha to show a clear structure of streets and blocks, as one would intuitively expect to see, but the density of dropoffs still seems approximately the same on nearly all Manhattan streets (just wider in some  ...
     "\n",
+    "Moreover, we're still ignoring 99% of the data.  Many plotting programs will have trouble with plots even this large, but Bokeh can handle 100-200,000 points in most browsers. Here we've enabled Bokeh's WebGL support, which gives smoother zooming behavior, but the non-WebGL mode also works well.\n",
```
OK. I'll re-add that line then.

Azaya89 commented on nyc_taxi/nyc_taxi.ipynb:
```diff
@@ -329,23 +320,41 @@
    "source": [
-    "When you zoom in interactively to this plot, you can see all the points available in that viewport without ever needing to change the plot parameters for that specific zoom level.  Each time you zoom or pan, a new image is rendered by Datashader (which takes a few seconds for large datasets), and displayed over or under other plot elements, providing full access to all of your data. Here we' ...
-    "\n",
+    "When you zoom in interactively to this plot, you can see all the points available in that viewport without ever needing to change the plot parameters for that specific zoom level. This new image is rendered using the `rasterize` parameter which uses Datashader to render the data into what is by default a 2D histogram, where every array cell counts the data points falling into that pixel. Bok ...
```
What I did here was basically copying over the explanation from the datashader link I shared.

maximlt replied:
@Azaya89 I suggest this text to replace the text between the plot and `# Customizing datashader`.

```
When you zoom in interactively to this plot, you can see all the points available in that viewport without ever needing to change the plot parameters for that specific zoom level. We called hvPlot with 3 parameters that enable and control the interactive Datashader pipeline orchestrated by HoloViews:

- `rasterize=True` is truly the only argument required to enable rasterization with Datashader. In this mode, a new 2D array is computed on every zoom and pan event on the plot, making use of all the points available in the viewport. By default, every array cell counts the number of data points falling into that pixel. This 2D array is then colormapped client-side by Bokeh (in your browser in Javascript, not in Python) applying the histogram equalization scheme we have already covered.
- With `aggregator=ds.sum('passenger_count')`, we configured the aggregation step of the pipeline to deviate from the default count of data points per pixel. Instead, it computes the total number of passengers per pixel. You can remove the specified `aggregator` if you prefer the default trip counts.
- We've also set the optional `dynspread` parameter to `True` to automatically enlarge the size of each data point once you've zoomed in so far that data points no longer have nearby neighbors.

Check this page out to learn more about the [principles of Datashading](https://holoviews.org/user_guide/Large_Data.html#principles-of-datashading).

Note that if you are viewing this plot as a static HTML exported web page, zooming in will just make the pixels larger; you need the full running Python process to have it re-rendered interactively.
```

Azaya89 commented:
> I updated your original post with the modernizing checklist.

Do we have to add this checklist to every new PR going forward? 'Cos usually I just look at it from the Google doc just before creating a new PR.

Azaya89 commented:
I have effected most of the requested changes. Pls let me know if I missed anything while we wait for the new holoviews version to be released :)

maximlt commented:
@Azaya89 hvPlot 0.10.0 has been released so you can now update the code to use `bgcolor` directly when calling hvPlot instead of via `.opts()`. When done, please ask Jim if he wants to review.

maximlt commented:
Merging, thanks @Azaya89, great job!
